inst/include/child_model_simulation.hpp

```diff
@@ -867,57 +867,68 @@ void run_child_natural_history(int t,
+      for (int cat = 0; cat < ss_c.hcTT; ++cat) {
+        for (int hd = 0; hd < ss_c.hc1DS; ++hd) {
+          auto hiv_deaths_per_group = n_hc.ctx_mean * n_hc.hc1_hiv_pop(hd, cat, a, s) * p_hc.hc1_cd4_mort(hd, cat, a);
+          i_hc.hc_posthivmort(hd, cat, a, s) = n_hc.hc1_hiv_pop(hd, cat, a, s) - hiv_deaths_per_group;
         }
       }
     }
   }
 
-  for (int s = 0; s < ss.NS; ++s) {
-    for (int a = hc_ss.hc2_agestart; a < pars.base.options.p_idx_fertility_first; ++a) {
-      for (int cat = 0; cat < hc_ss.hcTT; ++cat) {
-        for (int hd = 0; hd < hc_ss.hc2DS; ++hd) {
-          intermediate.children.hc_posthivmort(hd, cat, a, s) = state_next.children.hc2_hiv_pop(hd, cat, a - hc_ss.hc2_agestart, s) -
-            state_next.children.ctx_mean * state_next.children.hc2_hiv_pop(hd, cat, a - hc_ss.hc2_agestart, s) * cpars.hc2_cd4_mort(hd, cat, a - hc_ss.hc2_agestart);
+  for (int s = 0; s < ss_b.NS; ++s) {
+    for (int a = ss_c.hc2_agestart; a < p_op.p_idx_fertility_first; ++a) {
+      for (int cat = 0; cat < ss_c.hcTT; ++cat) {
+        for (int hd = 0; hd < ss_c.hc2DS; ++hd) {
+          auto hiv_deaths_per_group = n_hc.ctx_mean * n_hc.hc2_hiv_pop(hd, cat, a - ss_c.hc2_agestart, s) *
+                                      p_hc.hc2_cd4_mort(hd, cat, a - ss_c.hc2_agestart);
+          i_hc.hc_posthivmort(hd, cat, a, s) = n_hc.hc2_hiv_pop(hd, cat, a - ss_c.hc2_agestart, s) -
+                                               hiv_deaths_per_group;
         }
       }
     }
   }
 
-  //progress through CD4 categories
-  for (int s = 0; s < ss.NS; ++s) {
-    for (int hd = 1; hd < hc_ss.hc1DS; ++hd) {
-      for (int a = 0; a < hc_ss.hc2_agestart; ++a) {
-        for (int cat = 0; cat < hc_ss.hcTT; ++cat) {
-          intermediate.children.hc_grad(hd - 1, cat, a, s) -=  (intermediate.children.hc_posthivmort(hd - 1, cat, a, s) * cpars.hc1_cd4_prog(hd - 1, (cpars.hc_age_coarse_cd4(a)), s) + state_next.children.hc1_hiv_pop(hd - 1, cat, a, s) * cpars.hc1_cd4_prog(hd - 1, (cpars.hc_age_coarse_cd4(a)), s)) / 2; //moving to next cd4 category
-          intermediate.children.hc_grad(hd, cat, a, s) += (intermediate.children.hc_posthivmort(hd - 1, cat, a, s) * cpars.hc1_cd4_prog(hd - 1, (cpars.hc_age_coarse_cd4(a)), s) + state_next.children.hc1_hiv_pop(hd - 1, cat, a, s) * cpars.hc1_cd4_prog(hd - 1, (cpars.hc_age_coarse_cd4(a)), s)) / 2; //moving into this cd4 category
+  // progress through CD4 categories
+  for (int s = 0; s < ss_b.NS; ++s) {
+    for (int hd = 1; hd < ss_c.hc1DS; ++hd) {
+      for (int a = 0; a < ss_c.hc2_agestart; ++a) {
+        for (int cat = 0; cat < ss_c.hcTT; ++cat) {
+          const auto& coarse_hc1_cd4_prog = p_hc.hc1_cd4_prog(hd - 1, p_hc.hc_age_coarse_cd4(a), s);
+          auto hc_moving_to_curr_cd4 = coarse_hc1_cd4_prog *
+                                       (i_hc.hc_posthivmort(hd - 1, cat, a, s) + n_hc.hc1_hiv_pop(hd - 1, cat, a, s)) /
+                                       2.0;
```

Comment:
new variables, i think if we use `tensor(tensor())` for indexing in calculation then we should probably pull it out, its very confusing in the middle of a maths expression